query/worker.go

```diff
@@ -0,0 +1,247 @@
+package query
+
+import (
+	"errors"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+var (
+	// ErrQueryTimeout is an error returned if the worker doesn't respond
+	// with a valid response to the request within the timeout.
+	ErrQueryTimeout = errors.New("did not get response before timeout")
+
+	// ErrPeerDisconnected is returned if the worker's peer disconnect
+	// before the query has been answered.
+	ErrPeerDisconnected = errors.New("peer disconnected")
+
+	// ErrJobCanceled is returned if the job is canceled before the query
+	// has been answered.
+	ErrJobCanceled = errors.New("job canceled")
+)
+
+// queryJob is the internal struct that wraps the Query to work on, in
+// addition to some information about the query.
+type queryJob struct {
+	index      uint64
+	timeout    time.Duration
+	encoding   wire.MessageEncoding
+	cancelChan <-chan struct{}
+	*Request
+}
+
+// queryJob should satisfy the Task interface in order to be sorted by the
+// workQueue.
+var _ Task = (*queryJob)(nil)
+
+// Index returns the queryJob's index within the work queue.
+//
+// NOTE: Part of the Task interface.
+func (q *queryJob) Index() uint64 {
+	return q.index
+}
+
+// jobResult is the final result of the worker's handling of the queryJob.
+type jobResult struct {
+	job  *queryJob
+	peer Peer
+	err  error
+}
+
+// worker is responsible for polling work from its work queue, and handing it
+// to the associated peer. It validates incoming responses with the current
+// query's response handler, and polls more work for the peer when it has
+// successfully received a response to the request.
+type worker struct {
+	peer Peer
+
+	// nextJob is a channel of queries to be distributed, where the worker
+	// will poll new work from.
+	nextJob chan *queryJob
+}
+
+// NewWorker creates a new worker assosiated with the given peer.
+func NewWorker(peer Peer) *worker {
+	return &worker{
+		peer:    peer,
+		nextJob: make(chan *queryJob),
+	}
+}
+
+// Run starts the worker. The worker will supply its with queries, and handle
```

Comment:
Nit: it's peer*

Reply:
Fixed: https://github.com/lightninglabs/neutrino/pull/179/commits/c0f835f89d40a0250ad4094dec399364be94a201